File: simulator/abilities/on_hit_sneak_attack.py
from ..abilities.on_hit_effect import OnHit
from ..battle_map import Map
import numba_functions as nf
import logging

from ..utils.roll_types import RollType

logger = logging.getLogger("Encounterra")


class OnHitSneakAttack(OnHit):

    @staticmethod
    def get_dmg_dice(level):
        match level:
            case lvl if 1 <= lvl <= 2:
                return (1, 6)
            case lvl if 3 <= lvl <= 4:
                return (2, 6)
            case lvl if 5 <= lvl <= 6:
                return (3, 6)
            case lvl if 7 <= lvl <= 8:
                return (4, 6)
            case lvl if 9 <= lvl <= 10:
                return (5, 6)
            case lvl if 11 <= lvl <= 12:
                return (6, 6)
            case lvl if 13 <= lvl <= 14:
                return (7, 6)
            case lvl if 15 <= lvl <= 16:
                return (8, 6)
            case lvl if 17 <= lvl <= 18:
                return (9, 6)
            case lvl if 19 <= lvl <= 20:
                return (10, 6)
            case _:
                logger.error("Incorrect caster level of Sneak Attack")
                return (1, 6)

    def __init__(self, dmg_dice, dmg_type, crit_range, name="Sneak Attack"):
        self.dmg_dice = dmg_dice
        self.dmg_type = dmg_type
        self.crit_range = crit_range
        self.name = name

    def hit(self, attacker, attack, target, multiplier, dmg_so_far):
        if attack.roll_type is RollType.DISADVANTAGE:
            return None
        battle_map = Map.get()
        if not getattr(attacker, "already_used_sneak_attack_this_turn", True) and (attack.roll_type is RollType.ADVANTAGE or battle_map.is_ally_adjacent_to_target(attacker, target)):
            logger.info("Activating Sneak Attack")
            attacker.already_used_sneak_attack_this_turn = True
            return [nf.roll_dice_multi(self.dmg_dice) * multiplier, self.dmg_type]
        return None

    def calculate_threat(self, attacker, target, **kwargs):
        roll_type = kwargs.get("roll_type", RollType.STRAIGHT)
        if roll_type is RollType.DISADVANTAGE:
            return 0
        battle_map = Map.get()
        if not getattr(attacker, "already_used_sneak_attack_this_turn", True) and (roll_type is RollType.ADVANTAGE or battle_map.is_ally_adjacent_to_target(attacker, target)):
            avg_dmg_roll = nf.avg_roll_multi(self.dmg_dice)
            return avg_dmg_roll + 0.05 * self.crit_range * avg_dmg_roll
        return 0

File: c++/abilities/on_hit_sneak_attack.cpp
#include "abilities/on_hit_sneak_attack.hpp"
#include "actions/attack.hpp"
#include "core/battle_map.hpp"
#include "core/combatant.hpp"
#include <iostream>

namespace enc
{
  Die OnHitSneakAttack::getDmgDice(int level)
  {
    // 1d6 at levels 1-2, 2d6 at 3-4, 3d6 at 5-6, ... = ceil(level / 2) d6.
    int numDice = (level + 1) / 2;
    if(numDice < 1)
      {
        numDice = 1;
      }
    return Die{static_cast<unsigned char>(numDice), 6};
  }

  std::vector<std::pair<int, DamageType>>
  OnHitSneakAttack::hit(Combatant *attacker, Actoid *attack, Combatant *target, double multiplier, double dmgSoFar)
  {
    if(attacker == nullptr || target == nullptr)
      {
        return {};
      }

    // A rogue with Disadvantage on the attack roll cannot land a Sneak Attack.
    RollType rollType = RollType::STRAIGHT;
    if(auto *concreteAttack = dynamic_cast<Attack *>(attack))
      {
        rollType = concreteAttack->getRollType();
      }
    if(rollType == RollType::DISADVANTAGE)
      {
        return {};
      }

    if(attacker->hasAlreadyUsedSneakAttackThisTurn())
      {
        return {};
      }

    // Sneak Attack triggers with Advantage on the roll, or when an ally of the rogue is within 5 ft of the target.
    bool allyAdjacent = BattleMap::getInstance().isAllyAdjacentToTarget(*attacker, *target);
    if(rollType != RollType::ADVANTAGE && !allyAdjacent)
      {
        return {};
      }

    attacker->setAlreadyUsedSneakAttackThisTurn(true);
    int dmg = rollDiceMulti(_dmgDice);
    if(multiplier >= 2)
      {
        dmg *= 2;
      }
    std::cout << attacker->_name << " lands a Sneak Attack on " << target->_name << " for " << dmg << " damage" << std::endl;
    return {{dmg, _dmgType}};
  }

  double OnHitSneakAttack::calculateThreat(Combatant *attacker, Combatant *target, RollType rollType)
  {
    if(attacker == nullptr || target == nullptr)
      {
        return 0.0;
      }
    // Mirror Python OnHitSneakAttack.calculate_threat. Disadvantage disables Sneak Attack outright.
    if(rollType == RollType::DISADVANTAGE)
      {
        return 0.0;
      }
    if(attacker->hasAlreadyUsedSneakAttackThisTurn())
      {
        return 0.0;
      }
    // Sneak Attack is credited only when the rogue has Advantage on the roll (e.g. after Hide) or an ally is within
    // 5 ft of the target. The base attack-threat path calls this with a STRAIGHT roll, so a lone rogue's melee
    // attack is not over-valued by ~3d6 (which would wrongly out-score a genuine Disengage/Dash flee); the
    // Advantage credit is supplied by Hide::calculateThreatForAttack passing RollType::ADVANTAGE.
    if(rollType != RollType::ADVANTAGE && !BattleMap::getInstance().isAllyAdjacentToTarget(*attacker, *target))
      {
        return 0.0;
      }
    double avg = avgRollMulti(_dmgDice);
    return avg + 0.05 * _critRange * avg;
  }
}
